drivers/sensor/pixart/paa3905/paa3905.c

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (c) 2025 Croxel Inc.
+ * Copyright (c) 2025 CogniPilot Foundation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT pixart_paa3905
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/rtio/rtio.h>
+#include <zephyr/rtio/work.h>
+
+#include "paa3905.h"
+#include "paa3905_bus.h"
+#include "paa3905_stream.h"
+#include "paa3905_decoder.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(PAA3905, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Helper struct used to set reg-val sequences for op-modes */
+struct reg_val_pair {
+	uint8_t reg;
+	uint8_t val;
+};
+
+static void paa3905_submit_one_shot(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const struct sensor_chan_spec *const channels = cfg->channels;
+	const size_t num_channels = cfg->count;
+	uint32_t min_buf_len = sizeof(struct paa3905_encoded_data);
+	int err;
+	uint8_t *buf;
+	uint32_t buf_len;
+	struct paa3905_encoded_data *edata;
+
+	err = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (err) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, err);
+		return;
+	}
+
+	edata = (struct paa3905_encoded_data *)buf;
+
+	err = paa3905_encode(dev, channels, num_channels, buf);
+	if (err != 0) {
+		LOG_ERR("Failed to encode sensor data");
+		rtio_iodev_sqe_err(iodev_sqe, err);
+		return;
+	}
+
+	err = paa3905_bus_read(dev, REG_BURST_READ, edata->buf, sizeof(edata->buf));
+	if (err) {
+		LOG_ERR("Failed to read flow data");
+		rtio_iodev_sqe_err(iodev_sqe, err);
+		return;
+	}
+
+	LOG_HEXDUMP_INF(edata->buf, sizeof(edata->buf), "Flow data");
+
+	LOG_DBG("One shot reading complete");
+	rtio_iodev_sqe_ok(iodev_sqe, 0);
+}
+
+static void paa3905_submit_sync(struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const struct device *dev = cfg->sensor;
+
+	if (!cfg->is_streaming) {
+		paa3905_submit_one_shot(dev, iodev_sqe);
+	} else if (IS_ENABLED(CONFIG_PAA3905_STREAM)) {
+		paa3905_stream_submit(dev, iodev_sqe);
+	} else {
+		LOG_ERR("Streaming not supported");
+		rtio_iodev_sqe_err(iodev_sqe, -ENOTSUP);
+	}
+}
+
+static void paa3905_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	struct rtio_work_req *req = rtio_work_req_alloc();
```

Comment:
Why not do this asynchronously and avoid the work queue requirement altogether?

Reply:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/86644#discussion_r1987863462 resolution